x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -208,47 +219,118 @@ static void getMemoryAndProcessors(
                         estimatedMemoryUsage
                     )
                 );
-
-                extraSingleNodeModelMemoryInBytes = Math.max(extraSingleNodeModelMemoryInBytes, estimatedMemoryUsage);
+                extraPerNodeModelMemoryBytes = Math.max(extraPerNodeModelMemoryBytes, estimatedMemoryUsage);
                 extraModelMemoryInBytes += estimatedMemoryUsage;
 
-                // if not low priority, check processor requirements
+                // if not low priority, check processor requirements. If
                 if (Priority.LOW.equals(modelAssignment.getValue().getTaskParams().getPriority()) == false) {
-                    // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
-                    extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
-                    extraProcessors += numberOfAllocations * numberOfThreadsPerAllocation;
+                    if (numProcessorsNeeded > numberOfAvailableProcessors) {
+                        // as assignments can be placed on different nodes, we only need numberOfThreadsPerAllocation here
+                        extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
+                        extraProcessors += numberOfRequestedAllocations * numberOfThreadsPerAllocation - numExistingProcessorsToBeUsed;
+                    }
+                    processorsNewlyAssignedToTrainedModelsAllocationsThisTime += numExistingProcessorsToBeUsed;
                 }
             } else if (assignment.getNodeRoutingTable().values().stream().allMatch(r -> r.getState().consumesMemory() == false)) {
-                // Ignore states that don't consume memory, for example all allocations are failed
+                // Ignore states that don't consume memory, for example all allocations are failed or stopped
                 continue;
+            } else if (AssignmentState.STARTED.equals(assignment.getAssignmentState()) && numMissingAllocations > 0) {
+                // this state occurs when we have a partial allocation, but not fully allocated
+
+                long numMemoryBytesNeeded = assignment.getTaskParams().estimateMemoryUsageBytes();
+
+                if (numberOfAvailableProcessors < numProcessorsNeeded || perNodeAvailableModelMemoryBytes < numMemoryBytesNeeded) {
+                    // we don't have enough processors or memory to start the new allocations, we need to scale up
+
+                    processorsNewlyAssignedToTrainedModelsAllocationsThisTime += numExistingProcessorsToBeUsed;
+
+                    extraProcessors += numMissingAllocations * numberOfThreadsPerAllocation - numExistingProcessorsToBeUsed;
+                    extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, numberOfThreadsPerAllocation);
+                    extraModelMemoryInBytes += numMemoryBytesNeeded;
```

Comment:
How does this memory calculation work?

I think if the allocations go to the same node, they share the memory, but if they go to different nodes the each uses the model memory? Is that correct? 

Note that above (in STARTING) there's no memory computation.

I think adding `ceil(extraProcessorsIncrement / extraSingleNodeProcessors) * numMemoryBytesNeeded` is better if we care about memory here.

Reply:
You're raising a very good point here. We are not handling memory usage correctly/efficiently. 

These memory calculations assume that each deployment requires one unit of model memory, regardless of how it is actually deployed to nodes. 

The autoscaler assumes that model memory can be split accross nodes arbitrarily, rather than in discrete units, and calculates a total number of nodes that we need based on the total memory consumption and RAM of each node.

This is all problematic for multiple reasons:

- This algorithm doesn't consider model memory packing
- The autoscaler doesn't consider model memory packing
- Deploying a model with many allocations will require the model memory on every node it is deployed to, which could total a much larger number than we are suggesting here
- When deploying the same model in multiple deployments, this algorithm will require multiples of the model memory on each node, even if that is not required by the pytorch process
- Deploying multiple models without accounting for memory packing could make it impossible to actually pack the models onto any of the available nodes due to the above

Reply:
I think your idea of doing an estimation of the memory requirement based on the number of nodes makes sense, but I think with the packing issue we still set ouselves up to fail in corner cases. I think if we want to make this work we need to directly learn where the models have actually been placed. We may even need to try a hypothetical placement here before providing that scaling info to the autoscaler.

I think we would need to provide the memory and processors requirements for each node rather than just the sums. If we did this, we could also support heterogenous nodes to some degree.

This would be a significant overhaul, but its hard for me to see how we can avoid this.

Reply:
Fake availability zones have an impact on how these calculations work